# Proyek SQL 

Virus corona yang kehadirannya mengejutkan seluruh dunia telah mengubah rutinitas harian semua orang. Kini, warga kota tidak lagi menghabiskan waktu luangnya di luar rumah seperti pergi ke kafe dan mal. Mereka lebih sering berada di rumah, menghabiskan waktunya dengan membaca buku. Hal ini pun mendorong perusahaan startup untuk mengembangkan aplikasi baru bagi para pecinta buku.
<br>
# Tujuan
Karena meningkatnya minat baca buku, jumlah orang yang menggunakan aplikasi buku pun meningkat. Tujuan kami adalah untuk menghasilkan informasi yang cukup untuk memberikan dasar untuk membangun produk baru yang terkait dengan bidang aplikasi buku.
<br>
# Details
Database tersebut berisi data tentang buku, penerbit, penulis, serta rating dan ulasan pelanggan atas buku terkait. Informasi ini akan digunakan dalam membuat penawaran harga untuk sebuah produk baru.
<br>

__Deskripsi Data__
<br>
a. books:
<br>Berisi data tentang buku:
   - book_id — ID buku
   - author_id — ID penulis
   - title — judul buku
   - num_pages — jumlah halaman
   - publication_date — tanggal penerbitan
   - publisher_id — ID penerbit

<br>b. authors:
<br>Berisi data tentang penulis:
   - author_id — ID penulis
   - author — nama penulis

<br>c. publishers:
<br>Berisi data tentang penerbit:
   - publisher_id — ID penerbit
   - publisher — nama penerbit

<br>d. ratings:
<br>Berisi data tentang ulasan pengguna:
   - rating_id — ID rating
   - book_id — ID buku
   - username — nama pengguna yang memberi rating buku
   - rating

<br>e. reviews:
<br>Berisi data tentang ulasan pelanggan:
   - review_id — ID ulasan
   - book_id — ID buku
   - username — nama pengguna yang mengulas buku
   - text — teks ulasan
<br>

# Tugas
1. Hitung jumlah buku yang dirilis setelah tanggal 1 Januari 2000.
2. Hitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.
3. Identifikasi penerbit yang telah menerbitkan jumlah buku terbanyak, dengan jumlah halaman lebih dari 50 (hal ini akan membantu dalam mengecualikan brosur dan publikasi serupa dari analisis).
4. Identifikasi penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.
5. Hitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku.
<br>

# Instruksi 
1. Jelaskan tujuan dari studi yang dilakukan.
2. Pelajari tabelnya (tampilkan baris-baris pertamanya).
3. Buatlah kueri SQL untuk setiap tugas.
4. Tampilkan hasil dari setiap kueri tersebut pada notebook.
5. Kesimpulan

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>


In [1]:
# impor library yang diperlukan
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'practicum_student',         # nama pengguna
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # kata sandi
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # port koneksi
             'db': 'data-analyst-final-project-db'}          # nama database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
pd.io.sql.read_sql('select * from books limit 5', con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
pd.io.sql.read_sql('select * from authors limit 5', con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
pd.io.sql.read_sql('select * from ratings limit 5', con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [5]:
pd.io.sql.read_sql('select * from reviews limit 5', con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [6]:
pd.io.sql.read_sql('select * from publishers limit 5', con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>


# Hitung jumlah buku yang dirilis setelah tanggal 1 Januari 2000.

In [7]:
pd.io.sql.read_sql("select COUNT(DISTINCT book_id) AS cnt from books where publication_date < '2000-01-01'", 
                   con = engine)

cnt
0  179

In [8]:
pd.io.sql.read_sql("select COUNT(DISTINCT book_id) AS cnt from books where publication_date > '2000-01-01'", 
                   con = engine)

cnt
0  819

819 buku yang diterbitkan setelah tanggal 1 Januari 2000. Sekitar 4.5x lebih banyak jika dibandingkan dengan sebelum 1 January 2000.

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>


# Hitung jumlah ulasan pengguna dan rating rata-rata untuk setiap buku.

In [9]:
pd.io.sql.read_sql('select books.title, AVG(ratings.rating) AS average_rating,\
                   COUNT(DISTINCT reviews.review_id)\
                   AS cnt_reviews from ratings LEFT OUTER JOIN reviews ON ratings\
                   .username=reviews.username and ratings\
                   .book_id=reviews.book_id LEFT OUTER JOIN books ON ratings\
                   .book_id=books.book_id group by books.title', con = engine)


title  average_rating  \
0                                         'Salem's Lot        3.666667   
1                   1 000 Places to See Before You Die        2.500000   
2    13 Little Blue Envelopes (Little Blue Envelope...        4.666667   
3    1491: New Revelations of the Americas Before C...        4.500000   
4                                                 1776        4.000000   
..                                                 ...             ...   
994           Wyrd Sisters (Discworld  #6; Witches #2)        3.666667   
995                        Xenocide (Ender's Saga  #3)        3.400000   
996                                    Year of Wonders        3.200000   
997                        You Suck (A Love Story  #2)        4.500000   
998  Zen and the Art of Motorcycle Maintenance: An ...        3.833333   

     cnt_reviews  
0              2  
1              1  
2              3  
3              2  
4              4  
..           ...  
994            3  
995            3  
996            4  
997            2  
998            4  

[999 rows x 3 columns]

Untuk setiap buku, terlihat jumlah ulasan (cnt_reviews) dan peringkat rata-rata (average_rating).

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>


# Identifikasi penerbit yang telah menerbitkan jumlah buku terbanyak, dengan jumlah halaman lebih dari 50 (hal ini akan membantu dalam mengecualikan brosur dan publikasi serupa dari analisis).

In [9]:
pd.io.sql.read_sql("select publishers.publisher,\
COUNT(DISTINCT books.book_id)\
AS number_of_books from  books LEFT OUTER JOIN publishers ON books\
.publisher_id=publishers.publisher_id WHERE books.num_pages>50\
GROUP BY publishers.publisher ORDER BY number_of_books DESC LIMIT 1", con = engine)


publisher  number_of_books
0  Penguin Books               42

"Penguin Books" telah merilis jumlah buku terbanyak dengan lebih dari 50 halaman (42 buku)

In [10]:
pd.io.sql.read_sql(
        "select publishers.publisher_id AS publisher_id,\
        publishers.publisher AS publisher,\
        COUNT(DISTINCT(books.book_id)) AS number_of_books\
        FROM publishers\
        INNER JOIN books ON books.publisher_id = publishers.publisher_id\
        WHERE books.num_pages > 50\
        GROUP BY publishers.publisher_id, publishers.publisher\
        ORDER BY number_of_books DESC\
        LIMIT 20", con = engine)        



publisher_id                  publisher  number_of_books
0            212              Penguin Books               42
1            309                    Vintage               31
2            116   Grand Central Publishing               25
3            217           Penguin Classics               24
4             35                     Bantam               19
5             33           Ballantine Books               19
6             45                    Berkley               17
7             46              Berkley Books               14
8            284         St. Martin's Press               14
9            333  William Morrow Paperbacks               13
10            83                      Delta               13
11           129           Harper Perennial               12
12           133              HarperCollins               12
13           176  Little  Brown and Company               12
14           297                  Tor Books               11
15            31             Back Bay Books               11
16           143                HarperTorch               10
17           229               Pocket Books               10
18           210                    Penguin               10
19           178              Mariner Books               10

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>


# Identifikasi penulis dengan rating rata-rata buku tertinggi: temukan buku dengan rating minimal 50.

In [11]:
pd.io.sql.read_sql(
    "select Sub.author, AVG(Sub.average_rating)\
    AS average_rating_author from (select authors.author, ratings.book_id, AVG(ratings.rating)\
    AS average_rating, COUNT(DISTINCT ratings.rating_id)\
    AS cnt_rating from ratings LEFT OUTER JOIN books ON ratings.book_id=books.book_id LEFT OUTER JOIN authors\
    ON books.author_id=authors.author_id GROUP BY authors.author, ratings.book_id\
    HAVING COUNT(ratings.rating_id)>50 ORDER BY AVG(ratings.rating) DESC)\
    AS Sub group by author ORDER BY AVG(Sub.average_rating) DESC limit 1", con = engine)


author  average_rating_author
0  J.K. Rowling/Mary GrandPré               4.283844

Penulis dengan peringkat buku rata-rata tertinggi (buku dengan setidaknya 50 peringkat) adalah J.K. Rowling / Mary GrandPré

In [12]:
pd.io.sql.read_sql('''
    SELECT    
        books.author_id AS author_id,
        books.title AS title, 
        authors.author AS author_name,
        ratings.book_id AS book_id,
        AVG(ratings.rating) AS avg_rating    
    FROM
        ratings
    INNER JOIN
        books ON books.book_id = ratings.book_id
    INNER JOIN
        authors ON authors.author_id = books.author_id 
    GROUP BY
        ratings.book_id, books.author_id, books.title, authors.author
    HAVING
        COUNT(ratings.rating_id) > 49
    ORDER BY
        avg_rating DESC
    LIMIT 20;

''', con = engine)        

author_id                                              title  \
0         236  Harry Potter and the Prisoner of Azkaban (Harr...   
1         240  The Fellowship of the Ring (The Lord of the Ri...   
2         236  Harry Potter and the Chamber of Secrets (Harry...   
3         402                                     The Book Thief   
4         236  Harry Potter and the Half-Blood Prince (Harry ...   
5         376                                       Little Women   
6         236  Harry Potter and the Order of the Phoenix (Har...   
7         240                The Hobbit  or There and Back Again   
8         498  The Lightning Thief (Percy Jackson and the Oly...   
9         621                                  Lord of the Flies   
10        106             The Da Vinci Code (Robert Langdon  #2)   
11        235                             The Catcher in the Rye   
12        469                                      The Alchemist   
13        630                                   Romeo and Juliet   
14        372                          The Giver (The Giver  #1)   
15        195                                        Animal Farm   
16        106               Angels & Demons (Robert Langdon  #1)   
17        554                            Twilight (Twilight  #1)   
18        311                                    Of Mice and Men   

                                          author_name  book_id  avg_rating  
0                          J.K. Rowling/Mary GrandPré      302    4.414634  
1                                      J.R.R. Tolkien      722    4.391892  
2                          J.K. Rowling/Mary GrandPré      299    4.287500  
3                   Markus Zusak/Cao Xuân Việt Khương      656    4.264151  
4                          J.K. Rowling/Mary GrandPré      300    4.246575  
5                                   Louisa May Alcott      399    4.192308  
6                          J.K. Rowling/Mary GrandPré      301    4.186667  
7                                      J.R.R. Tolkien      750    4.125000  
8                                        Rick Riordan      779    4.080645  
9                                     William Golding      405    3.901408  
10                                          Dan Brown      696    3.830508  
11                                      J.D. Salinger      673    3.825581  
12           Paulo Coelho/Alan R. Clarke/Özdemir İnce      627    3.789474  
13  William Shakespeare/Paul Werstine/Barbara A. M...      545    3.787879  
14                                         Lois Lowry      733    3.750000  
15            George Orwell/Boris Grabnar/Peter Škerl       79    3.729730  
16                                          Dan Brown       75    3.678571  
17                                    Stephenie Meyer      948    3.662500  
18                                     John Steinbeck      488    3.622951

Tiga buku dengan average rating tertinggi adalah:
1. Buku dari J.K. Rowling/Mary GrandPré dengan judul "Harry Potter and..."
2. Buku dari J.R.R Tolkien dengan judul "The Fellowship of The Ring..."
3. Buku dari J.K. Rowling/Mary GrandPré dengan judul "Harry Potter and..."

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>


# Hitung jumlah rata-rata teks ulasan di antara pengguna yang memberi rating terhadap lebih dari 50 buku.

In [13]:
pd.io.sql.read_sql(
    'select AVG(Sub.cnt_reviews)\
    AS average from (select ratings.username,COUNT(DISTINCT ratings.book_id)\
    AS cnt_rating, COUNT(DISTINCT reviews.review_id)\
    AS cnt_reviews from ratings LEFT OUTER JOIN reviews\
    ON ratings.username=reviews.username and ratings.book_id=reviews.book_id group by ratings.username\
    having COUNT(DISTINCT ratings.book_id)>50) AS Sub', con = engine)


average
0  24.333333

Jumlah rata-rata ulasan teks di antara pengguna yang menilai lebih dari 50 buku adalah 24 ulasan

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>


# Kesimpulan
1. Jumlah buku yang dirilis setelah 1 January 2000 adalah 819, sedangkan 179 buku dirilis sebelum 1 January 2000. Minat baca memang tumbuh pesat > 1 January 2000
2. Penerbit yang paling aktif menerbitkan buku adalah Penguin Books, Vintage, Grand Central Publishing dan Penguin Classics. 
3. Penulis buku / author paling laris manis (karena mendapat rating tinggi) adalah J.K. Rowling/Mary GrandPré dengan judul bukunya mengenai Harry Potter. 
4. Rata-rata ulasan (text) dari penggunan (untuk kategori > 50 buku) adalah 24 ulasan. 
5. Kesimpulan ini mengarah pada kecenderungan positif bagi industri buku terutama dengan genre seperti yang ditulis oleh J.K. Rowling/Mary GrandPré dan J.R.R. Tolkien setelah 1 January 2000. 

<div class="alert alert-success">
<b>Code Reviewers's comment v.1</b> <a class="tocSkip"></a>

Bagus, semua berjalan lancar.

</div>
